File: Lecture_Example/week1/tail.py
def count(n):
    if n == 0:
        return 0
    else:
        return 1 + count(n-1)


def count_tail(n, acc):
    if n == 0:
        return acc
    else:
        return count_tail(n - 1, acc + 1)


def count_iterative(n, acc):
    while n != 0:
        n, acc = n - 1, acc + 1

    return acc

File: Lecture_Example/week4/Tail.hs
{-# LANGUAGE BangPatterns #-}

count :: Int -> Int
count n =
    if n == 0
    then
        0
    else
        1 + count (n - 1)


-- Try this with and without the exclamation mark in front of acc.
countTail :: Int -> Int -> Int
countTail n !acc =
    if n == 0
    then
        acc
    else
        countTail (n - 1) (acc + 1)


main :: IO ()
main = print (countTail 15000000 0)
